### PR TITLE
Open external links in Safari

### LINKED ISSUE #2517: Open external links in Safari
It would be handy to have a setting to open all external links (or indeed GitHub links that GitHawk don’t handle through native UI) in Safari instead of in-app WebView.
This is a setting that you can find in many iOS apps and no doubt would be useful addition to many users.

### PR DESCRIPTION
**Pull request by ismetanin**
Fixes #2517 

**Changes:**
Added setting for external links opening in Safari

**Notes:**
I'm not sure that `func presentSafari(url: URL)` is a proper name for our method. And also I have some doubts about `Defaults+ExternalLinks.swift` implementation

### CODE REVIEW

File: Classes/Settings/SettingsViewController.swift [hunk 1]
```diff
@@ -46,6 +46,8 @@ GitHubSessionListener {
     @IBOutlet weak var pushSwitch: UISwitch!
     @IBOutlet weak var pushCell: UITableViewCell!
     @IBOutlet weak var pushSettingsButton: UIButton!
+    @IBOutlet weak var openExternalLinksLabel: SettingsLabel!
```
Comment (Huddie) [inline Classes/Settings/SettingsViewController.swift]: Is this needed? Do you use this variable anywhere?

File: Classes/Settings/Defaults+ExternalLinks.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+//  Copyright © 2018 Ryan Nystrom. All rights reserved.
+//
+
+import Foundation
+
+extension UserDefaults {
+
+    private static let defaultKey = "com.whoisryannystrom.freetime.is-need-open-external-links-in-safari"
+
+    var isNeedOpenExternalLinksInSafari: Bool {
```
Comment (Huddie) [inline Classes/Settings/Defaults+ExternalLinks.swift]: Maybe “shouldOpenExternal...” isNeed sounds a bit off to me.

**Review by Huddie — COMMENTED** [review]
What will happen before any user defaults are set (before the user switches the settings updating the user default value)? Will it be true/false/nil?

**Comment by ismetanin** [comment]
@Huddie by default bool value in UserDefaults in **false**. 

> Return Value
 The Boolean value associated with the specified key. If the specified key doesn‘t exist, this method returns false.

Details see [here](https://developer.apple.com/documentation/foundation/userdefaults/1416388-bool).

File: Classes/Settings/SettingsViewController.swift (hunk 1 above)
Comment (ismetanin) [reply Classes/Settings/SettingsViewController.swift]: Nope
Maybe it will be needed later for localization. But okay, now I remove it

**Comment by Huddie** [comment]
@ismetanin cool 😎! 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Huddie** [comment]
@ismetanin you added a file but didn’t upload the project .pbxproj file. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ismetanin** [comment]
@Huddie fixed

File: Classes/Settings/Defaults+ExternalLinks.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+//
+//  Created by Ivan Smetanin on 08/12/2018.
+//  Copyright © 2018 Ryan Nystrom. All rights reserved.
+//
+
+import Foundation
+
+extension UserDefaults {
+
+    private static let defaultKey = "com.whoisryannystrom.freetime.is-need-open-external-links-in-safari"
```
Comment (BasThomas) [inline Classes/Settings/Defaults+ExternalLinks.swift]: Why don't we name this `should-open-external-links-in-safari` too? Not that big of a deal of course, just wondering why.

File: Classes/View Controllers/UIViewController+Safari.swift [hunk 4]
```diff
@@ -12,8 +12,12 @@ import SafariServices
 extension UIViewController {
 
     func presentSafari(url: URL) {
-		guard let safariViewController = try? SFSafariViewController.configured(with: url) else { return }
-        route_present(to: safariViewController)
+        if UserDefaults.standard.shouldOpenExternalLinksInSafari {
+            UIApplication.shared.open(url)
```
Comment (BasThomas) [inline Classes/View Controllers/UIViewController+Safari.swift]: Does this not need a `canOpenURL(_ url:)` check?

File: Classes/View Controllers/UIViewController+Safari.swift [hunk 5]
```diff
@@ -12,8 +12,12 @@ import SafariServices
...
+        } else {
+            guard let safariViewController = try? SFSafariViewController.configured(with: url) else { return }
```
Comment (BasThomas) [inline Classes/View Controllers/UIViewController+Safari.swift]: What about putting this in a `do/catch` and asserting if this fails?

File: Classes/View Controllers/UIViewController+Safari.swift [hunk 6]
```diff
@@ -12,8 +12,12 @@ import SafariServices
...
+            route_present(to: safariViewController)
```
Comment (BasThomas) [inline Classes/View Controllers/UIViewController+Safari.swift]: @rnystrom: why is this using snake case? 😮

File: Classes/View Controllers/UIViewController+Safari.swift (hunk 5 above)
Comment (ismetanin) [reply Classes/View Controllers/UIViewController+Safari.swift]: Good suggestion!

File: Classes/Settings/Defaults+ExternalLinks.swift (hunk 3 above)
Comment (ismetanin) [reply Classes/Settings/Defaults+ExternalLinks.swift]: Forgot to fix

File: Classes/View Controllers/UIViewController+Safari.swift (hunk 4 above)
Comment (ismetanin) [reply Classes/View Controllers/UIViewController+Safari.swift]: It won't crash but in general, it would be good to check
Maybe put assertion here too?

Comment (BasThomas) [reply Classes/View Controllers/UIViewController+Safari.swift]: Sounds good

File: Classes/View Controllers/UIViewController+Safari.swift (hunk 6 above)
Comment (rnystrom) [reply Classes/View Controllers/UIViewController+Safari.swift]: experimenting w/ prefix for a utils extension

Comment (ismetanin) [reply Classes/View Controllers/UIViewController+Safari.swift]: @rnystrom Should I fix it?

Comment (rnystrom) [reply Classes/View Controllers/UIViewController+Safari.swift]: Not a big deal right now IMO

**Comment by rnystrom** [comment]
Love this. Has been requested multiple times!

**Review by CryptoGodfatherVA — COMMENTED** [review]
#

**Pull request merged**